CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/visual/DefaultInteractionEventHandler.java

```diff
@@ -1099,23 +1101,46 @@ private void showContextMenu(final GraphReadMethods rg, final Camera camera, fin
         final Vector3f graphLocation = visualInteraction.windowToGraphCoordinates(camera, screenLocation);
 
         for (final ContextMenuProvider pmp : popups) {
+            // Retrive list of item names to populate and optional list of icons to assign
+            // Icons will be added corresponding to an image if the icons list is not ull
+            // and a non null icon is provided with corresponding index.
             final List<String> items = pmp.getItems(rg, elementType, clickedId);
+            final List<ImageIcon> icons = pmp.getIcons(rg, elementType, clickedId);
             if (!items.isEmpty()) {
                 final List<String> menuPath = pmp.getMenuPath(elementType);
                 if (CollectionUtils.isEmpty(menuPath)) {
-                    items.forEach(item -> {
-                        popup.add(new AbstractAction(item) {
-                            @Override
-                            public void actionPerformed(final ActionEvent event) {
-                                PluginExecution.withPlugin(new SimplePlugin("Context Menu: " + item) {
-                                    @Override
-                                    protected void execute(final PluginGraphs _graphs, final PluginInteraction _interaction, final PluginParameters _parameters) throws InterruptedException, PluginException {
-                                        pmp.selectItem(item, graph, elementType, clickedId, graphLocation);
-                                    }
-                                }).executeLater(null);
-                            }
-                        });
-                    });
+                    for (int idx = 0; idx < items.size(); idx++) {
+                        Icon icon = null;
+                        if (icons != null && icons.size() > idx + 1) {  icon = (Icon)icons.get(idx); }
```

Comment:
These 2 lines could be combined into a ternary statement (icon could then be made final)